### DISCUSSION TITLE
Keyboard + mouse to move/resize window from anywhere inside it

### DISCUSSION THREAD

**Discussion by jab** [rxhanson/Rectangle#1009] [Ideas]
Sometimes users need to move or resize a window ad hoc (e.g. "I just need to nudge this a bit down and left") rather than to one of the pre-calculated precise tiled locations.

I've found that the fastest way to accomplish this is to hold down a modifier key while mousing from anywhere within the window, à la [Hummingbird.app](https://github.com/finestructure/Hummingbird#readme) (not to mention the many Linux window managers like KDE that support this natively).

Would you consider adding support for this to Rectangle? If Rectangle added this, it would have all my window management needs covered in a single app.

Thanks for your consideration and for your work on this great app.

**Comment by rxhanson** [comment]
Thanks for the feedback. Rectangle is intentionally slim, and I don’t have plans to add it in. 

This behavior is in [Rectangle Pro](https://rectangleapp.com/pro) (paid).

**Reply by jab** [reply]
I don't see this in Rectangle Pro. I only see how to "throw" a window to a precise pre-calculated tiled size and position, not a size/position that changes gradually as the mouse is moved. Am I missing something or was there a misunderstanding of what this is asking for? Thanks for the quick reply.

**Reply by rxhanson** [reply]
On the window throw tab in the prefs window it is move and resize, and you need to record your desired modifier keys. (I’m away from my Mac but can provide a screenshot when I get back to it if needed)

**Reply by jab** [reply]
I see now, thank you. Looks like this isn't enabled by default, but is easy to configure. I think I missed this because I assumed the "window throw" section was all about the snapping behavior, and I didn't see the "move" and "resize" sections at the bottom. I wonder if it'd be worth pulling them out into their own section (or otherwise making them more discoverable), and ideally even enabling shortcuts for them out-of-the-box.

In any case, great to know this is possible with Rectangle Pro. Now that I've got it working, I can report a minor bug (see screenshot below): I think the error message has a typo in the phrase "conflict with what you have set for Reticle", and would ideally name the specific conflict, in this case the conflict was with the main "Window Throw" trigger.

<img width="889" alt="Screen Shot 2022-11-26 at 6 18 20 PM" src="https://user-images.githubusercontent.com/64992/204112378-e8f4f7df-2a23-4ce4-a35c-fb6ffcf08da6.png">

Also, for full move/resize feature parity with Hummingbird, there should be an option to not focus the window being moved/resized via the shortcut, but rather allow it to be moved/resized in the background.

Hope this helps!

**Reply by rxhanson** [reply]
Thanks for the additional feedback. 

> I see now, thank you. Looks like this isn't enabled by default, but is easy to configure. I think I missed this because I assumed the "window throw" section was all about the snapping behavior, and I didn't see the "move" and "resize" sections at the bottom. I wonder if it'd be worth pulling them out into their own section (or otherwise making them more discoverable), and ideally even enabling shortcuts for them out-of-the-box.

This used to be enabled by default, but I pulled it at a certain point. There are a lot of features in the app, and having things enabled by default leads to unexpected behavior that might not be desired. Not all people want this, and I have to be selective in what is enabled by default or else it would be quite annoying to have to go into an app and disable a bunch of stuff that can be considered intrusive upon install or have to play a guessing game to find out what things might be causing some unexpected behavior. The window throw has the initial window that explains the configured window throw modifiers, and I find the window throw to be the most useful feature and that's why it has a default (without the window throw, I would have never made Rectangle Pro). Also, it appears that a lot of users that enable move/resize don't like the same modifiers. You would have changed away from the modifiers I used to have as the defaults. I agree that it is confusing to have those in the Window Throw tab (a leftover relic of when the app had far fewer features). I don't like adding more tabs since there are already a lot, but I'll give this some more thought on making them more discoverable as you mentioned.

> In any case, great to know this is possible with Rectangle Pro. Now that I've got it working, I can report a minor bug (see screenshot below): I think the error message has a typo in the phrase "conflict with what you have set for Reticle", and would ideally name the specific conflict, in this case the conflict was with the main "Window Throw" trigger.

Thanks! I'll fix it.

> Also, for full move/resize feature parity with Hummingbird, there should be an option to not focus the window being moved/resized via the shortcut, but rather allow it to be moved/resized in the background.

You can disable the automatic 'bring to front' for move/resize with the following terminal command followed by an app restart:

`defaults write com.knollsoft.Hookshot windowToFront -int 1`

(The default value is 0)

This would of course be better with UI config and/or some dynamic toggling. It's on my backlog already but hasn't quite bubbled up to the top.

**Reply by jab** [reply]
Thanks, `defaults write com.knollsoft.Hookshot windowToFront -int 1` worked, and nice a checkbox for that is on the to-do list.
